examples/coap-server.c

```diff
@@ -341,13 +343,54 @@ init_resources(coap_context_t *ctx) {
 #endif /* WITHOUT_ASYNC */
 }
 
+static int
+verify_cn_callback(const char *cn,
+                   const uint8_t *asn1_public_cert UNUSED_PARAM,
+                   size_t asn1_length UNUSED_PARAM,
+                   coap_session_t *session UNUSED_PARAM,
+                   unsigned depth,
+                   int validated UNUSED_PARAM,
+                   void *arg UNUSED_PARAM
+) {
+  coap_log(LOG_INFO, "CN '%s' presented by client (%s)\n",
+           cn, depth ? "CA" : "Certificate");
+  return 1;
+}
+
 static void
 fill_keystore(coap_context_t *ctx) {
   if (cert_file) {
     coap_dtls_pki_t dtls_pki;
     memset (&dtls_pki, 0, sizeof(dtls_pki));
-    dtls_pki.public_cert = cert_file;
-    dtls_pki.private_key = cert_file;
+    dtls_pki.version = COAP_DTLS_PKI_SETUP_VERSION;
+    if (ca_file) {
+      /*
+       * Add in additional certificate checking.
+       * This list of enabled can be tuned for the specific
+       * requirements - see 'man coap_encryption'.
+       */
+      dtls_pki.verify_peer_cert        = 1;
+      dtls_pki.require_peer_cert       = 1;
+      dtls_pki.allow_self_signed       = 1;
+      dtls_pki.allow_expired_certs     = 1;
+      dtls_pki.cert_chain_validation   = 1;
+      dtls_pki.cert_chain_verify_depth = 2;
+      dtls_pki.check_cert_revocation   = 1;
+      dtls_pki.allow_no_crl            = 1;
+      dtls_pki.allow_expired_crl       = 1;
+      dtls_pki.validate_cn_call_back   = verify_cn_callback;
+      dtls_pki.cn_call_back_arg        = NULL;
+      dtls_pki.validate_sni_call_back  = NULL;
+      dtls_pki.sni_call_back_arg       = NULL;
+    }
+    dtls_pki.pki_key.key_type = COAP_PKI_KEY_PEM;
+    dtls_pki.pki_key.key.pem.public_cert = cert_file;
+    dtls_pki.pki_key.key.pem.private_key = cert_file;
+    dtls_pki.pki_key.key.pem.ca_file = ca_file;
+    /* If general root CAs are defined */
+    if (root_ca_file) {
+      coap_context_set_pki_root_cas(ctx, root_ca_file, NULL);
```

Comment:
During testing, I found it very useful to specify a directory such as `/etc/ssl/certs` like this:

```
diff --git a/examples/coap-server.c b/examples/coap-server.c
index 5f9d2c7..5f75009 100644
--- a/examples/coap-server.c
+++ b/examples/coap-server.c
@@ -389,7 +389,12 @@ fill_keystore(coap_context_t *ctx) {
     dtls_pki.pki_key.key.pem.ca_file = ca_file;
     /* If general root CAs are defined */
     if (root_ca_file) {
-      coap_context_set_pki_root_cas(ctx, root_ca_file, NULL);
+      struct stat stbuf;
+      if ((stat(root_ca_file, &stbuf) == 0) && S_ISDIR(stbuf.st_mode)) {
+        coap_context_set_pki_root_cas(ctx, NULL, root_ca_file);
+      } else {
+        coap_context_set_pki_root_cas(ctx, root_ca_file, NULL);
+      }
     }
     coap_context_set_pki(ctx, &dtls_pki);
     if (key_defined)
```

Reply:
Yes, I like that change.